# Wrangle Pisa 2012
#### By Gabriela Sikora


### Abstract



### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [3]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [4]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

/Users/gabriela/Desktop/Code/myenv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [5]:
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [6]:
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [7]:
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [8]:
# Check for null values in relevant columns
columns = ['STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [9]:
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [10]:
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Make the variable names readable since they are codes that can only be read with the dictionary
3. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')
5. Create a total for Out-of-School Study Time and Learning time (minutes per week)
6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [11]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### Narrow down the dataframe by only looking at variables of interest

There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [12]:
pisa_clean = pisa_clean[['STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 'ST28Q01', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [13]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST28Q01',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### Create an average of all Math, Reading, and Science scores

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [14]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score'] +pisa_clean['Average Math Score']) / 3

In [15]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [16]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
count    485490.000000
mean        629.039175
std         128.839689
min         105.496473
25%         536.312167
50%         627.318087
75%         721.099990
max        1123.882273
Name: Average Total Score, dtype: float64


In [17]:
pisa_clean.sample(10)

STIDSTD ST04Q01                ST13Q01 ST14Q01 ST14Q02 ST14Q03  \
24284      2134  Female      <ISCED level 3A>       No     Yes      No   
316972     1270  Female  <ISCED level 3B, 3C>       No      No      No   
377890     5004  Female      <ISCED level 3A>      NaN     NaN     NaN   
471669     1432    Male      <ISCED level 3A>       No      No      No   
411743      996    Male      <ISCED level 3A>      NaN     NaN     Yes   
245555     4966  Female      <ISCED level 3A>       No      No      No   
345189    29487    Male      <ISCED level 3A>       No      No      No   
212227      657    Male  <ISCED level 3B, 3C>       No      No     Yes   
364233      324    Male                    NaN      No      No      No   
171269      374    Male      <ISCED level 3A>       No      No     Yes   

       ST14Q04                ST17Q01 ST18Q01 ST18Q02  ... ST57Q03 ST57Q04  \
24284      Yes      <ISCED level 3A>       No     Yes  ...     0.0     0.0   
316972     NaN       <ISCED level 2>       No      No  ...     NaN     NaN   
377890     Yes                    NaN     NaN     NaN  ...     0.0     0.0   
471669     NaN  <ISCED level 3B, 3C>       No      No  ...     NaN     NaN   
411743     Yes      <ISCED level 3A>       No      No  ...     NaN     NaN   
245555      No      <ISCED level 3A>       No      No  ...     NaN     NaN   
345189     NaN  <ISCED level 3B, 3C>      NaN     Yes  ...     NaN     NaN   
212227     NaN      <ISCED level 3A>       No      No  ...     0.0     0.0   
364233      No                    NaN      No      No  ...     NaN     NaN   
171269      No      <ISCED level 3A>       No      No  ...     0.0     0.0   

       ST57Q05  MMINS  LMINS  SMINS  Average Math Score  \
24284      0.0  180.0  270.0  135.0           426.32028   
316972     NaN    NaN    NaN    NaN           355.43698   
377890     0.0    NaN    NaN    NaN           325.99314   
471669     NaN    NaN    NaN    NaN           492.45208   
411743     NaN    NaN    NaN    NaN           550.48296   
245555     NaN  200.0  350.0  200.0           530.38636   
345189     NaN    NaN    NaN    NaN           537.24100   
212227     0.0  180.0  135.0  180.0           517.92336   
364233     NaN    NaN    NaN    NaN           519.63702   
171269     0.0  225.0  135.0  135.0           583.58780   

        Average Reading Score  Average Science Score  Average Total Score  
24284               477.14602              438.95934           589.581973  
316972              367.61068              356.24766           478.244100  
377890              358.55552              319.32114           443.287647  
471669              463.50602              465.06900           637.826393  
411743              539.61056              553.65534           731.410607  
245555              532.98604              535.09884           709.619200  
345189              601.76134              520.55202           732.265120  
212227              462.62384              568.38866           688.953073  
364233              520.36390              548.06042           702.566120  
171269              560.62154              581.16374           769.653627  

[10 rows x 25 columns]

### Make the variable names readable

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [18]:
# Rename columns
pisa_clean.rename({'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST28Q01': 'How many books at home', 
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning time (minutes per week)- <Mathematics>',
                   'LMINS': 'Learning time (minutes per week)  - <test language>',
                   'SMINS': 'Learning time (minutes per week) - <Science>'}, axis='columns', inplace=True)

In [19]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

### Education
Multiple columns with the single variable. 

http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

In [319]:
father_edu = pisa_clean.copy()

In [320]:
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [321]:
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [322]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [323]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [324]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
288763        3712  Female      <ISCED level 3B, 3C>    
181234        5560    Male          <ISCED level 3A>    
342770       27068  Female          <ISCED level 3A>    
417517        5047  Female           <ISCED level 2>    
212176         606    Male          <ISCED level 3A>    
239666        4132  Female          <ISCED level 3A>    
47451         6065    Male      <ISCED level 3B, 3C>    
147540        1958    Male           <ISCED level 1>    
479968        4416  Female          <ISCED level 3A>    
284123        5423  Female          <ISCED level 3A>    

       Mother Qualifications - ISCED level 6  \
288763                                    No   
181234                                   NaN   
342770                                    No   
417517                                    No   
212176                                    No   
239666                                    No   
47451                                    Yes   
147540                                    No   
479968                                    No   
284123                                    No   

       Mother Qualifications - ISCED level 5A  \
288763                                     No   
181234                                    NaN   
342770                                     No   
417517                                     No   
212176                                     No   
239666                                     No   
47451                                     NaN   
147540                                     No   
479968                                     No   
284123                                     No   

       Mother Qualifications - ISCED level 5B  \
288763                                    Yes   
181234                                    Yes   
342770                                    Yes   
417517                                     No   
212176                                    Yes   
239666                                     No   
47451                                     NaN   
147540                                     No   
479968                                    Yes   
284123                                    Yes   

       Mother Qualifications - ISCED level 4 Father - Highest Schooling  \
288763                                    No          <ISCED level 3A>    
181234                                   NaN      <ISCED level 3B, 3C>    
342770                                   NaN                        NaN   
417517                                    No                        NaN   
212176                                    No          <ISCED level 3A>    
239666                                    No      <ISCED level 3B, 3C>    
47451                                    NaN           <ISCED level 2>    
147540                                    No           <ISCED level 1>    
479968                                    No                        NaN   
284123                                   NaN          <ISCED level 3A>    

       <ISCED level 6>  <ISCED level 5>  ...  \
288763              No               No  ...   
181234             NaN              NaN  ...   
342770             NaN  <ISCED level 5>  ...   
417517              No               No  ...   
212176              No               No  ...   
239666              No               No  ...   
47451               No               No  ...   
147540              No               No  ...   
479968             NaN              NaN  ...   
284123              No               No  ...   

       Learning time (minutes per week) - <Science> Average Math Score  \
288763                                        585.0          421.10142   
181234                                        120.0          559.67440   
342770                                        200.0          372.80730   
417517                                          NaN          451.16840   
212176                                        

In [325]:
father_edu = pd.melt(father_edu, 
                     id_vars=['Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'How many books at home',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning time (minutes per week)- <Mathematics>',
                            'Learning time (minutes per week)  - <test language>',
                            'Learning time (minutes per week) - <Science>',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [326]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
3083885          50  Female          <ISCED level 3A>    
4144558       20049    Male                        NaN   
4330851        3337  Female      <ISCED level 3B, 3C>    
1841243        1245    Male           <ISCED level 2>    
3803523        4877    Male          <ISCED level 3A>    
432308         3987    Male      <ISCED level 3B, 3C>    
4757811        4873  Female           <ISCED level 2>    
2063370        7312    Male           <ISCED level 2>    
1308152       21470    Male          <ISCED level 3A>    
1550786       19847    Male          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
3083885                                    No   
4144558                                   NaN   
4330851                                    No   
1841243                                    No   
3803523                                    No   
432308                                    NaN   
4757811                                    No   
2063370                                    No   
1308152                                    No   
1550786                                    No   

        Mother Qualifications - ISCED level 5A  \
3083885                                     No   
4144558                                    NaN   
4330851                                     No   
1841243                                     No   
3803523                                    Yes   
432308                                     NaN   
4757811                                     No   
2063370                                     No   
1308152                                    Yes   
1550786                                     No   

        Mother Qualifications - ISCED level 5B  \
3083885                                     No   
4144558                                    NaN   
4330851                                     No   
1841243                                     No   
3803523                                     No   
432308                                     Yes   
4757811                                     No   
2063370                                     No   
1308152                                     No   
1550786                                    Yes   

        Mother Qualifications - ISCED level 4 How many books at home  \
3083885                                    No         101-200 books    
4144558                                   NaN         201-500 books    
4330851                                   NaN         101-200 books    
1841243                                    No            0-10 books    
3803523                                   Yes         201-500 books    
432308                                    NaN          26-100 books    
4757811                                    No            0-10 books    
2063370                                   NaN            0-10 books    
1308152                                   NaN           11-25 books    
1550786                                   Yes           11-25 books    

         Out-of-School Study Time - Homework  \
3083885                                  NaN   
4144558                                  NaN   
4330851                                 12.0   
1841243                                  1.0   
3803523                                  NaN   
432308                                   NaN   
4757811                                  5.0   
2063370                                 12.0   
1308152                                  2.0   
1550786                                  0.0   

         Out-of-School Study Time - Guided Homework  ...  \
3083885                                         NaN  ...   
4144558                                         NaN  ...   
4330851                                         0.0  ...   
1841243                                         1.0  ...   
3803523                                         NaN  ...   
432308                                        

In [327]:
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [328]:
father_edu.sample(70)

Student ID  Gender             Mother - Highest Schooling  \
2355636        1206  Female                       <ISCED level 2>    
415295         2825  Female                                    NaN   
3675901        5809    Male                  <ISCED level 3B, 3C>    
3677239         109  Female                      <ISCED level 3A>    
2288970       31308  Female                       <ISCED level 1>    
315390         5023    Male                      <ISCED level 3A>    
1741393        6223  Female                      <ISCED level 3A>    
7065           2323  Female                       <ISCED level 2>    
1481734        3114    Male                                    NaN   
699033         1973  Female                      <ISCED level 3A>    
4079551        6515  Female                  <ISCED level 3B, 3C>    
3917076       11006    Male                      <ISCED level 3A>    
58686          3421  Female                  <ISCED level 3B, 3C>    
314587         4220    Male                       <ISCED level 2>    
167470        21888  Female                      <ISCED level 3A>    
2948040       12950  Female                  <ISCED level 3B, 3C>    
3508972        3300    Male                      <ISCED level 3A>    
2035341       18912    Male                      <ISCED level 3A>    
421134         3590    Male                      <ISCED level 3A>    
1858221        1535    Male                      <ISCED level 3A>    
142353         4252    Male                      <ISCED level 3A>    
2278852       21190  Female                       <ISCED level 2>    
125615         2444  Female                       <ISCED level 2>    
1463894        2682  Female                      <ISCED level 3A>    
3292844         983  Female                       <ISCED level 2>    
1360021        5513  Female                       <ISCED level 2>    
742740        16661  Female                      <ISCED level 3A>    
2207481       24932    Male                      <ISCED level 3A>    
2948015       12925  Female                       <ISCED level 2>    
197100         7984  Female                  <ISCED level 3B, 3C>    
...             ...     ...                                    ...   
3181650       28121  Female                      <ISCED level 3A>    
247758         7169    Male                       <ISCED level 2>    
3945654        6469  Female                       <ISCED level 2>    
3868782         115    Male                       <ISCED level 2>    
3228136        4829    Male                       <ISCED level 2>    
3617           3618  Female                       <ISCED level 2>    
4235189        1761  Female                  <ISCED level 3B, 3C>    
1877326        3312    Male                      <ISCED level 3A>    
19917          3675    Male                       <ISCED level 2>    
250327         9738  Female                       <ISCED level 1>    
3409959        6787    Male                      <ISCED level 3A>    
1552921         438    Male                       <ISCED level 2>    
3845292        3268  Female                      <ISCED level 3A>    
4083450       10414  Female                      <ISCED level 3A>    
210838         3938    Male                  <ISCED level 3B, 3C>    
1914574        3728  Female                      <ISCED level 3A>    
1808104        2126    Male                  <ISCED level 3B, 3C>    
3886916        2997    Male                  <ISCED level 3B, 3C>    
288943         3892    Male                       <ISCED level 2>    
3355300        4677    Male                  <ISCED level 3B, 3C>    
244715         4126    Male                       <ISCED level 2>    
4368927        4477  Female  She did not complete <ISCED level 1>    
2761865       18713    Male                       <ISCED level 2>    
4062815        3221    Male                  <ISCED level 3B, 3C>    
831928        30736    Male                       <ISCED level 2>    
1930193        3486

In [329]:
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [330]:
father_edu['Education - Father'].value_counts()

<ISCED level 3>    306459
<ISCED level 5>    185130
<ISCED level 2>     84329
<ISCED level 4>     77265
<ISCED level 1>     35938
<ISCED level 6>     29589
<ISCED level 0>     16535
Name: Education - Father, dtype: int64

In [331]:
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [332]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
1546012       15073  Female          <ISCED level 3A>    
1413195        4532  Female      <ISCED level 3B, 3C>    
1196690        4322  Female          <ISCED level 3A>    
1288038        1356    Male      <ISCED level 3B, 3C>    
3408319        5147    Male          <ISCED level 3A>    
2106261       18719  Female          <ISCED level 3A>    
1774980        2808    Male          <ISCED level 3A>    
1439094        2725  Female          <ISCED level 3A>    
3605231        5026  Female          <ISCED level 3A>    
1314182       27500    Male      <ISCED level 3B, 3C>    

        Mother Qualifications - ISCED level 6  \
1546012                                    No   
1413195                                   Yes   
1196690                                    No   
1288038                                    No   
3408319                                    No   
2106261                                    No   
1774980                                   NaN   
1439094                                    No   
3605231                                    No   
1314182                                    No   

        Mother Qualifications - ISCED level 5A  \
1546012                                    Yes   
1413195                                    NaN   
1196690                                    NaN   
1288038                                    Yes   
3408319                                    Yes   
2106261                                    Yes   
1774980                                    Yes   
1439094                                    Yes   
3605231                                     No   
1314182                                    Yes   

        Mother Qualifications - ISCED level 5B  \
1546012                                     No   
1413195                                    NaN   
1196690                                    NaN   
1288038                                    Yes   
3408319                                     No   
2106261                                    Yes   
1774980                                    NaN   
1439094                                     No   
3605231                                     No   
1314182                                    Yes   

        Mother Qualifications - ISCED level 4 How many books at home  \
1546012                                    No           11-25 books    
1413195                                   NaN         101-200 books    
1196690                                   NaN          26-100 books    
1288038                                   NaN    More than 500 books   
3408319                                    No          26-100 books    
2106261                                   Yes         201-500 books    
1774980                                   NaN         101-200 books    
1439094                                   NaN          26-100 books    
3605231                                   Yes          26-100 books    
1314182                                   NaN          26-100 books    

         Out-of-School Study Time - Homework  \
1546012                                 15.0   
1413195                                  2.0   
1196690                                  3.0   
1288038                                  NaN   
3408319                                 14.0   
2106261                                  4.0   
1774980                                 21.0   
1439094                                 10.0   
3605231                                 14.0   
1314182                                  NaN   

         Out-of-School Study Time - Guided Homework  ...  \
1546012                                         2.0  ...   
1413195                                         NaN  ...   
1196690                                         2.0  ...   
1288038                                         NaN  ...   
3408319                                         1.0  ...   
2106261                                       

### Mother Edu

In [ ]:
father_edu = pisa_clean.copy()

In [283]:
father_edu['Student ID'].duplicated().sum()

1714751

In [307]:
# https://stackoverflow.com/questions/20067264/pandas-drop-duplicates-with-condition
father_edu_groups = father_edu.groupby('Student ID')

In [308]:
father_edu_groups['Education - Father'].max().reset_index()

TypeError: '>=' not supported between instances of 'str' and 'float'

In [105]:
father_edu_groups.head()

Student ID  Gender             Mother - Highest Schooling  \
485492            3  Female                  <ISCED level 3B, 3C>    
485493            4  Female                  <ISCED level 3B, 3C>    
485500           11  Female                      <ISCED level 3A>    
485505           16    Male                       <ISCED level 2>    
485517           28    Male                      <ISCED level 3A>    
485535           46  Female                      <ISCED level 3A>    
485546           57  Female                      <ISCED level 3A>    
485558           69    Male                  <ISCED level 3B, 3C>    
485608          119  Female                      <ISCED level 3A>    
485622          133  Female                      <ISCED level 3A>    
485629          140    Male                      <ISCED level 3A>    
485668          179  Female                                    NaN   
485686          197    Male                  <ISCED level 3B, 3C>    
485740          251  Female                       <ISCED level 2>    
485756          267  Female                      <ISCED level 3A>    
485773          284  Female                  <ISCED level 3B, 3C>    
485792          303  Female                      <ISCED level 3A>    
485819          330    Male                      <ISCED level 3A>    
485843          354  Female                                    NaN   
485875          386  Female                      <ISCED level 3A>    
485890          401  Female                       <ISCED level 2>    
485905          416  Female                      <ISCED level 3A>    
485906          417  Female                      <ISCED level 3A>    
485920          431  Female                      <ISCED level 3A>    
485935          446  Female                  <ISCED level 3B, 3C>    
485951          462  Female                      <ISCED level 3A>    
485958          469    Male                      <ISCED level 3A>    
485963          474    Male                       <ISCED level 2>    
485981          492    Male                  <ISCED level 3B, 3C>    
485984          495  Female                      <ISCED level 3A>    
...             ...     ...                                    ...   
4718674       33562    Male  She did not complete <ISCED level 1>    
4718690       33578    Male                       <ISCED level 2>    
4718708       33596    Male                       <ISCED level 2>    
4718734       33622    Male                       <ISCED level 2>    
4718758       33646    Male                  <ISCED level 3B, 3C>    
4718797       33685    Male                       <ISCED level 1>    
4718799       33687  Female                                    NaN   
4718803       33691    Male                       <ISCED level 2>    
4718804       33692  Female                       <ISCED level 1>    
4718811       33699  Female  She did not complete <ISCED level 1>    
4718812       33700    Male  She did not complete <ISCED level 1>    
4718815       33703    Male                       <ISCED level 2>    
4718819       33707  Female                       <ISCED level 2>    
4718820       33708  Female                       <ISCED level 1>    
4718825       33713    Male                       <ISCED level 2>    
4718831       33719  Female  She did not complete <ISCED level 1>    
4718832       33720    Male  She did not complete <ISCED level 1>    
4718836       33724  Female                       <ISCED level 1>    
4718843       33731  Female                       <ISCED level 2>    
4718858       33746    Male                       <ISCED level 1>    
4718865       33753  Female                       <ISCED level 2>    
4718881       33769  Female  She did not complete <ISCED level 1>    
4718885       33773  Female                       <ISCED level 1>    
4718895       33783    Male                       <ISCED level 1>    
4718897       33785    Male                       <ISCED level 2>    
4718901       33789

In [109]:
father_edu_groups.duplicated

AttributeError: Cannot access callable attribute 'duplicated' of 'DataFrameGroupBy' objects, try using the 'apply' method

In [19]:
father_edu = pisa_clean.copy()

In [20]:
pisa_clean[ 'Father - Highest Schooling'].value_counts()

<ISCED level 3A>                        215280
<ISCED level 3B, 3C>                     91179
<ISCED level 2>                          84329
<ISCED level 1>                          35938
He did not complete <ISCED level 1>      16535
Name: Father - Highest Schooling, dtype: int64

In [21]:
father_edu.rename({'Father Qualifications - ISCED level 6': '<ISCED level 6>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>'}, axis='columns', inplace=True)


In [22]:
father_edu = pd.melt(father_edu, 
                     id_vars=['Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'Father - Highest Schooling', 'How many books at home',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning time (minutes per week)- <Mathematics>',
                            'Learning time (minutes per week)  - <test language>',
                            'Learning time (minutes per week) - <Science>',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='father_edu', value_name='father')



In [23]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
769883         5693  Female          <ISCED level 3A>    
450527         2197    Male      <ISCED level 3B, 3C>    
1502523        4667    Male                        NaN   
1253483        3803  Female          <ISCED level 3A>    
1176286        3531  Female                        NaN   
1337319        2430  Female      <ISCED level 3B, 3C>    
1011046        3435    Male          <ISCED level 3A>    
592074        10571  Female          <ISCED level 3A>    
852019         2620    Male          <ISCED level 3A>    
1052029        6580  Female          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
769883                                     No   
450527                                     No   
1502523                                    No   
1253483                                    No   
1176286                                   NaN   
1337319                                    No   
1011046                                    No   
592074                                    Yes   
852019                                     No   
1052029                                   Yes   

        Mother Qualifications - ISCED level 5A  \
769883                                      No   
450527                                      No   
1502523                                     No   
1253483                                     No   
1176286                                    NaN   
1337319                                     No   
1011046                                    Yes   
592074                                      No   
852019                                     Yes   
1052029                                    NaN   

        Mother Qualifications - ISCED level 5B  \
769883                                     NaN   
450527                                     Yes   
1502523                                     No   
1253483                                    NaN   
1176286                                    NaN   
1337319                                    Yes   
1011046                                     No   
592074                                      No   
852019                                     Yes   
1052029                                    Yes   

        Mother Qualifications - ISCED level 4 Father - Highest Schooling  \
769883                                    NaN          <ISCED level 3A>    
450527                                    NaN          <ISCED level 3A>    
1502523                                   NaN      <ISCED level 3B, 3C>    
1253483                                   NaN                        NaN   
1176286                                   NaN                        NaN   
1337319                                   Yes          <ISCED level 3A>    
1011046                                    No          <ISCED level 3A>    
592074                                     No          <ISCED level 3A>    
852019                                    Yes      <ISCED level 3B, 3C>    
1052029                                   NaN          <ISCED level 3A>    

        How many books at home  Out-of-School Study Time - Homework  ...  \
769883             0-10 books                                   NaN  ...   
450527     More than 500 books                                  NaN  ...   
1502523            0-10 books                                   2.0  ...   
1253483         101-200 books                                   1.0  ...   
1176286                    NaN                                  NaN  ...   
1337319          26-100 books                                   2.0  ...   
1011046    More than 500 books                                  2.0  ...   
592074     More than 500 books                                 20.0  ...   
852019            11-25 books                                   NaN  ...   
1052029         201-500 books                                   1.0  ...   

         Out-of-School Study Time - With Parent  \
769

In [31]:
list(father_edu)

['Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'father_edu',
 'father']

In [32]:
father_edu_miss = father_edu[(father_edu['father'] == 'No') | (father_edu['father'] == 'NaN')]

In [33]:
father_edu_miss

Student ID  Gender             Mother - Highest Schooling  \
1                 2  Female                      <ISCED level 3A>    
4                 5  Female  She did not complete <ISCED level 1>    
5                 6  Female                  <ISCED level 3B, 3C>    
7                 8    Male                      <ISCED level 3A>    
9                10  Female                      <ISCED level 3A>    
12               13  Female                       <ISCED level 2>    
13               14  Female                      <ISCED level 3A>    
14               15    Male                       <ISCED level 2>    
16               17  Female                       <ISCED level 2>    
17               18    Male                       <ISCED level 2>    
19               20  Female                      <ISCED level 3A>    
21               22    Male                       <ISCED level 2>    
22               23  Female                       <ISCED level 2>    
23               24  Female                       <ISCED level 2>    
24               25    Male                       <ISCED level 2>    
25               26  Female                       <ISCED level 2>    
26               27  Female                      <ISCED level 3A>    
28               29  Female                       <ISCED level 2>    
29               30  Female                      <ISCED level 3A>    
31               32    Male                       <ISCED level 1>    
33               34    Male                       <ISCED level 2>    
34               35    Male                      <ISCED level 3A>    
35               36    Male                      <ISCED level 3A>    
36               37  Female                       <ISCED level 2>    
37               38  Female                       <ISCED level 2>    
39               40  Female                      <ISCED level 3A>    
40               41    Male                      <ISCED level 3A>    
42               43  Female                      <ISCED level 3A>    
43               44    Male                  <ISCED level 3B, 3C>    
44               45  Female                       <ISCED level 2>    
...             ...     ...                                    ...   
1941926        4926    Male                       <ISCED level 2>    
1941927        4927    Male                       <ISCED level 1>    
1941928        4928    Male                       <ISCED level 1>    
1941929        4929    Male                      <ISCED level 3A>    
1941930        4930    Male                       <ISCED level 1>    
1941931        4931    Male  She did not complete <ISCED level 1>    
1941932        4932  Female                                    NaN   
1941934        4934    Male                       <ISCED level 2>    
1941935        4935    Male  She did not complete <ISCED level 1>    
1941936        4936    Male  She did not complete <ISCED level 1>    
1941937        4937  Female  She did not complete <ISCED level 1>    
1941938        4938  Female  She did not complete <ISCED level 1>    
1941939        4939  Female  She did not complete <ISCED level 1>    
1941940        4940    Male                       <ISCED level 1>    
1941941        4941    Male                       <ISCED level 2>    
1941942        4942    Male                       <ISCED level 1>    
1941943        4943  Female                       <ISCED level 1>    
1941944        4944    Male                       <ISCED level 2>    
1941945        4945    Male  She did not complete <ISCED level 1>    
1941946        4946  Female                       <ISCED level 2>    
1941947        4947  Female  She did not complete <ISCED level 1>    
1941948        4948  Female                       <ISCED level 1>    
1941949        4949    Male                       <ISCED level 1>    
1941953        4953  Female                      <ISCED level 3A>    
1941954        4954    Male  She did not complete <ISCED level 1>    
1941955        4955

In [34]:
father_edu_miss['father_edu'].value_counts()

<ISCED level 5>    565500
<ISCED level 6>    341826
<ISCED level 4>    214828
Name: father_edu, dtype: int64

In [35]:
father_edu_miss['father_edu'].replace('<ISCED level 6>', np.nan, inplace=True)
father_edu_miss['father_edu'].replace('<ISCED level 5>', np.nan, inplace=True)
father_edu_miss['father_edu'].replace('<ISCED level 4>', np.nan, inplace=True)

In [36]:
father_edu_miss.head()

Student ID  Gender             Mother - Highest Schooling  \
1           2  Female                      <ISCED level 3A>    
4           5  Female  She did not complete <ISCED level 1>    
5           6  Female                  <ISCED level 3B, 3C>    
7           8    Male                      <ISCED level 3A>    
9          10  Female                      <ISCED level 3A>    

  Mother Qualifications - ISCED level 6  \
1                                   Yes   
4                                    No   
5                                    No   
7                                    No   
9                                    No   

  Mother Qualifications - ISCED level 5A  \
1                                    Yes   
4                                     No   
5                                     No   
7                                    Yes   
9                                     No   

  Mother Qualifications - ISCED level 5B  \
1                                     No   
4                                     No   
5                                     No   
7                                     No   
9                                     No   

  Mother Qualifications - ISCED level 4 Father - Highest Schooling  \
1                                    No          <ISCED level 3A>    
4                                    No      <ISCED level 3B, 3C>    
5                                    No           <ISCED level 2>    
7                                    No          <ISCED level 3A>    
9                                    No          <ISCED level 3A>    

  How many books at home  Out-of-School Study Time - Homework  ...  \
1         201-500 books                                   NaN  ...   
4         101-200 books                                   NaN  ...   
5            0-10 books                                   3.0  ...   
7          26-100 books                                   NaN  ...   
9           11-25 books                                   NaN  ...   

   Out-of-School Study Time - With Parent  \
1                                     0.0   
4                                     NaN   
5                                     1.0   
7                                     NaN   
9                                     9.0   

   Learning time (minutes per week)- <Mathematics>  \
1                                            270.0   
4                                              NaN   
5                                              NaN   
7                                              NaN   
9                                             90.0   

   Learning time (minutes per week)  - <test language>  \
1                                              315.0     
4                                                NaN     
5                                                NaN     
7                                                NaN     
9                                              240.0     

   Learning time (minutes per week) - <Science>  Average Math Score  \
1                                          90.0           470.56396   
4                                           NaN           385.50398   
5                                           NaN           365.17370   
7                                           NaN           542.14830   
9                                         270.0           433.87600   

   Average Reading Score  Average Science Score  Average Total Score  \
1              384.68832              478.12382           601.313353   
4              256.01010              367.15778           464.725280   
5              337.42688              375.55018           481.108153   
7              516.59478              559.06378           719.985053   
9              401.92486              342.16710           537.281320   

   father_edu  father  
1         NaN      No  
4         NaN      No  
5         NaN      No  
7         NaN      No  
9         NaN      No  

[5 rows x 23 columns]

In [48]:
father_edu_clean = father_edu.merge(father_edu_miss, on='Student ID')
father_edu_clean.drop(columns=['father'], axis=1, inplace=True)

In [46]:
father_edu_miss.shape

(1122154, 23)

In [47]:
pisa_clean.shape

(485490, 25)

rename column father_edu to Father - Highest Schooling - then concat and drop nans 

### 4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [18]:

pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

In [19]:
# Check to see if null values correctly dropped
pisa_clean

Country  International Grade  Gender  \
3       Albania                    9  Female   
6       Albania                   10  Female   
13      Albania                   10  Female   
17      Albania                   10    Male   
18      Albania                   10  Female   
27      Albania                   10    Male   
28      Albania                   10  Female   
46      Albania                    9  Female   
53      Albania                   10    Male   
55      Albania                    9    Male   
57      Albania                    9    Male   
58      Albania                   10    Male   
61      Albania                   10  Female   
62      Albania                   10  Female   
63      Albania                    9  Female   
66      Albania                    9    Male   
73      Albania                    9  Female   
75      Albania                    9    Male   
84      Albania                   10  Female   
86      Albania                   10  Female   
87      Albania                   10    Male   
89      Albania                   10  Female   
92      Albania                   10    Male   
94      Albania                   10    Male   
96      Albania                    9    Male   
97      Albania                   10    Male   
103     Albania                   10  Female   
108     Albania                   10  Female   
123     Albania                   10    Male   
124     Albania                   10    Male   
...         ...                  ...     ...   
485378  Vietnam                   10  Female   
485383  Vietnam                   10    Male   
485393  Vietnam                   10    Male   
485398  Vietnam                   10  Female   
485399  Vietnam                   10  Female   
485400  Vietnam                   10  Female   
485407  Vietnam                   10    Male   
485411  Vietnam                   10    Male   
485412  Vietnam                   10    Male   
485417  Vietnam                   10    Male   
485419  Vietnam                   10  Female   
485424  Vietnam                   10  Female   
485425  Vietnam                   10    Male   
485427  Vietnam                   10  Female   
485429  Vietnam                   10    Male   
485431  Vietnam                   10  Female   
485434  Vietnam                   10  Female   
485435  Vietnam                   10    Male   
485437  Vietnam                   10  Female   
485440  Vietnam                   10    Male   
485442  Vietnam                   10    Male   
485444  Vietnam                   10  Female   
485446  Vietnam                   10  Female   
485453  Vietnam                   10  Female   
485458  Vietnam                   10    Male   
485460  Vietnam                   10    Male   
485463  Vietnam                   10    Male   
485469  Vietnam                   10  Female   
485472  Vietnam                   10    Male   
485475  Vietnam                   10    Male   

                    Mother<Highest Schooling>  \
3                       <ISCED level 3B, 3C>    
6                       <ISCED level 3B, 3C>    
13                          <ISCED level 3A>    
17                           <ISCED level 2>    
18                          <ISCED level 3A>    
27                          <ISCED level 3A>    
28                           <ISCED level 2>    
46                           <ISCED level 2>    
53                          <ISCED level 3A>    
55                           <ISCED level 2>    
57      She did not complete <ISCED level 1>    
58                           <ISCED level 2>    
61                          <ISCED level 3A>    
62                           <ISCED level 2>    
63                           <ISCED level 2>    
66                          <ISCED level 3A>    
73                          <ISCED level 3A>    
75                          <ISCED level 3A>    
84                           <ISCED level 2>    
86                          <ISCED level 3A>    
87

In [20]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

(189773, 18)

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [21]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning time (minutes per week) - Total'] = pisa_clean['Learning time (minutes per week)- <Mathematics>'] + pisa_clean['Learning time (minutes per week)  - <test language>'] + pisa_clean['Learning time (minutes per week) - <Science>'] 
                                                                                                                 

In [22]:
list(pisa_clean)

['Country',
 'International Grade',
 'Gender',
 'Mother<Highest Schooling>',
 'Father<Highest Schooling>',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score',
 'Out-of-School Study Time - Total',
 'Learning time (minutes per week) - Total']

### 6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)

In [23]:
pisa_clean.replace(['Connecticut (USA)', 'Florida (USA)', 'Massachusetts (USA)'], 'United States of America', inplace=True)
pisa_clean.replace(['Perm(Russian)', 'Russian Federation', 'Perm(Russian Federation)'], 'Russia', inplace=True)

In [24]:
pisa_clean['Country'].value_counts()

Italy                       14482
Mexico                      11510
Spain                       11149
Canada                      10773
United Kingdom               6046
Switzerland                  5383
United States of America     5279
Australia                    5150
Brazil                       4553
Finland                      4502
United Arab Emirates         3616
Belgium                      3491
Denmark                      3474
Japan                        3413
Thailand                     3161
Singapore                    3005
Chinese Taipei               2960
Croatia                      2827
Poland                       2774
China-Shanghai               2723
Czech Republic               2722
Korea                        2566
Russia                       2543
Estonia                      2510
Ireland                      2448
Slovenia                     2369
Macao-China                  2333
Sweden                       2310
Luxembourg                   2286
Netherlands   

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [25]:
pisa_clean

Country  International Grade  Gender  \
3       Albania                    9  Female   
6       Albania                   10  Female   
13      Albania                   10  Female   
17      Albania                   10    Male   
18      Albania                   10  Female   
27      Albania                   10    Male   
28      Albania                   10  Female   
46      Albania                    9  Female   
53      Albania                   10    Male   
55      Albania                    9    Male   
57      Albania                    9    Male   
58      Albania                   10    Male   
61      Albania                   10  Female   
62      Albania                   10  Female   
63      Albania                    9  Female   
66      Albania                    9    Male   
73      Albania                    9  Female   
75      Albania                    9    Male   
84      Albania                   10  Female   
86      Albania                   10  Female   
87      Albania                   10    Male   
89      Albania                   10  Female   
92      Albania                   10    Male   
94      Albania                   10    Male   
96      Albania                    9    Male   
97      Albania                   10    Male   
103     Albania                   10  Female   
108     Albania                   10  Female   
123     Albania                   10    Male   
124     Albania                   10    Male   
...         ...                  ...     ...   
485378  Vietnam                   10  Female   
485383  Vietnam                   10    Male   
485393  Vietnam                   10    Male   
485398  Vietnam                   10  Female   
485399  Vietnam                   10  Female   
485400  Vietnam                   10  Female   
485407  Vietnam                   10    Male   
485411  Vietnam                   10    Male   
485412  Vietnam                   10    Male   
485417  Vietnam                   10    Male   
485419  Vietnam                   10  Female   
485424  Vietnam                   10  Female   
485425  Vietnam                   10    Male   
485427  Vietnam                   10  Female   
485429  Vietnam                   10    Male   
485431  Vietnam                   10  Female   
485434  Vietnam                   10  Female   
485435  Vietnam                   10    Male   
485437  Vietnam                   10  Female   
485440  Vietnam                   10    Male   
485442  Vietnam                   10    Male   
485444  Vietnam                   10  Female   
485446  Vietnam                   10  Female   
485453  Vietnam                   10  Female   
485458  Vietnam                   10    Male   
485460  Vietnam                   10    Male   
485463  Vietnam                   10    Male   
485469  Vietnam                   10  Female   
485472  Vietnam                   10    Male   
485475  Vietnam                   10    Male   

                    Mother<Highest Schooling>  \
3                       <ISCED level 3B, 3C>    
6                       <ISCED level 3B, 3C>    
13                          <ISCED level 3A>    
17                           <ISCED level 2>    
18                          <ISCED level 3A>    
27                          <ISCED level 3A>    
28                           <ISCED level 2>    
46                           <ISCED level 2>    
53                          <ISCED level 3A>    
55                           <ISCED level 2>    
57      She did not complete <ISCED level 1>    
58                           <ISCED level 2>    
61                          <ISCED level 3A>    
62                           <ISCED level 2>    
63                           <ISCED level 2>    
66                          <ISCED level 3A>    
73                          <ISCED level 3A>    
75                          <ISCED level 3A>    
84                           <ISCED level 2>    
86                          <ISCED level 3A>    
87

In [26]:
pisa_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 189773 entries, 3 to 485475
Data columns (total 20 columns):
Country                                                189773 non-null object
International Grade                                    189773 non-null int64
Gender                                                 189773 non-null object
Mother<Highest Schooling>                              189773 non-null object
Father<Highest Schooling>                              189773 non-null object
How many books at home                                 189773 non-null object
Out-of-School Study Time - Homework                    189773 non-null float64
Out-of-School Study Time - Guided Homework             189773 non-null float64
Out-of-School Study Time - Personal Tutor              189773 non-null float64
Out-of-School Study Time - Commercial Company          189773 non-null float64
Out-of-School Study Time - With Parent                 189773 non-null float64
Learning time (minutes per 

<a id='sav'></a>
## Store

In [27]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References